apps/omg_eth/lib/omg_eth/ethereum_height_monitor.ex

```diff
@@ -0,0 +1,237 @@
+  # its internal state according to the raised alarms. Mainly these handlers do 2 things:
+  #
+  # 1. Reflect the internal state so that it does not re-raise an existing alarm
+  # 2. Maintain the list of events that the alarm corresponds to
+  #
+  def handle_cast({:set_alarm, :ethereum_connection_error}, state) do
+    events = [%Event.EthereumConnectionError{} | state.events]
+    {:noreply, %{state | connection_alarm_raised: true, events: events}}
+  end
+
+  def handle_cast({:clear_alarm, :ethereum_connection_error}, state) do
+    events = clear_events(state.events, Event.EthereumConnectionError)
+    {:noreply, %{state | connection_alarm_raised: false, events: events}}
+  end
+
+  def handle_cast({:set_alarm, :ethereum_stalled_sync}, state) do
+    events = [
+      %Event.EthereumStalledSync{
+        ethereum_height: state.ethereum_height,
+        synced_at: state.synced_at
+      }
+      | state.events
+    ]
+
+    {:noreply, %{state | stall_alarm_raised: true, events: events}}
+  end
+
+  def handle_cast({:clear_alarm, :ethereum_stalled_sync}, state) do
+    events = clear_events(state.events, Event.EthereumStalledSync)
+    {:noreply, %{state | stall_alarm_raised: false, events: events}}
+  end
+
+  #
+  # Private functions
+  #
+
+  @spec stalled?(non_neg_integer() | :error, non_neg_integer(), DateTime.t(), non_neg_integer()) :: boolean()
+  defp stalled?(height, previous_height, synced_at, stall_threshold_ms) do
+    case height do
+      height when is_integer(height) and height > previous_height ->
+        false
+
+      _ ->
+        DateTime.diff(DateTime.utc_now(), synced_at, :millisecond) > stall_threshold_ms
+    end
+  end
+
+  @spec fetch_height() :: non_neg_integer() | :error
+  defp fetch_height() do
+    case eth().get_ethereum_height() do
+      {:ok, height} ->
+        height
+
+      error ->
+        _ = Logger.warn("Error retrieving Ethereum height: #{inspect(error)}")
+        :error
+    end
+  end
+
+  @spec eth() :: module()
+  defp eth(), do: Application.get_env(:omg_eth, :eth_integration_module, OMG.Eth)
+
+  @spec broadcast_on_new_height(module(), non_neg_integer(), non_neg_integer() | :error) :: :ok | {:error, term()}
+  defp broadcast_on_new_height(_event_bus, _previous_height, :error), do: :ok
+
+  defp broadcast_on_new_height(event_bus, previous_height, height) when height > previous_height do
+    apply(event_bus, :broadcast, ["ethereum_new_height", {:ethereum_new_height, height}])
+  end
+
+  defp broadcast_on_new_height(_, _, _), do: :ok
+
+  @spec clear_events(events_t(), module()) :: events_t()
+  defp clear_events(events, module) do
+    Enum.reject(events, fn %struct_module{} -> struct_module == module end)
+  end
+
+  #
+  # Alarms management
+  #
+
+  defp install_alarm_handler do
+    case Enum.member?(:gen_event.which_handlers(:alarm_handler), __MODULE__.AlarmHandler) do
+      true -> :ok
+      _ -> :alarm_handler.add_alarm_handler(__MODULE__.AlarmHandler)
```

Comment:
if you'd like to parametrize AlarmHandler (idk, for testing?) you could use add_alarm_handler/2 https://github.com/yrashk/erlang/blob/master/lib/sasl/src/alarm_handler.erl#L67

Reply:
Let's do this when we do the byzantine events regrouping? I'm gonna need some time to work out this flow for async testing.